deployment/ccip/changeset/cs_rmn_curse_uncurse.go

```diff
@@ -0,0 +1,275 @@
+				return fmt.Errorf("chain %s: %w", targetChain.String(), err)
+			}
+
+			if !slices.Contains(validSelectors, action.ChainSelector) {
+				return errors.Errorf("invalid chain selector %d for chain %s", action.ChainSelector, targetChain.String())
+			}
+
+			if !slices.Contains(validSubjects, action.SubjectToCurse) {
+				return errors.Errorf("invalid subject %x for chain %s", action.SubjectToCurse, targetChain.String())
+			}
+		}
+	}
+
+	return nil
+}
+
+type Subject = [16]byte
+
+func SelectorToSubject(subject uint64) Subject {
+	var b Subject
+	binary.BigEndian.PutUint64(b[8:], subject)
+	return b
+}
+
+// CurseLaneOnlyOnSource curses a lane only on the source chain
+// This will prevent message from source to destination to be initiated
+// One noteworthy behaviour is that this means that message can be sent from destination to source but will not be executed on the source
+func CurseLaneOnlyOnSource(sourceSelector uint64, destinationSelector uint64) CurseAction {
+	// Curse from source to destination
+	return func(e deployment.Environment) []RMNCurseAction {
+		return []RMNCurseAction{
+			{
+				ChainSelector:  sourceSelector,
+				SubjectToCurse: SelectorToSubject(destinationSelector),
+			},
+		}
+	}
+}
+
+func CurseLane(sourceSelector uint64, destinationSelector uint64) CurseAction {
+	// Bidirectional curse between two chains
+	return func(e deployment.Environment) []RMNCurseAction {
+		return append(
+			CurseLaneOnlyOnSource(sourceSelector, destinationSelector)(e),
+			CurseLaneOnlyOnSource(destinationSelector, sourceSelector)(e)...,
+		)
+	}
+}
+
+func CurseChain(chainSelector uint64) CurseAction {
+	return func(e deployment.Environment) []RMNCurseAction {
+		chainSelectors := e.AllChainSelectors()
+
+		// Curse all other chains to prevent onramp from sending message to the cursed chain
+		var curseActions []RMNCurseAction
+		for _, otherChainSelector := range chainSelectors {
+			if otherChainSelector != chainSelector {
+				curseActions = append(curseActions, RMNCurseAction{
+					ChainSelector:  otherChainSelector,
+					SubjectToCurse: SelectorToSubject(chainSelector),
+				})
+			}
+		}
+
+		// Curse the chain with a global curse to prevent any onramp or offramp message from send message in and out of the chain
+		curseActions = append(curseActions, RMNCurseAction{
+			ChainSelector:  chainSelector,
+			SubjectToCurse: GlobalCurseSubject(),
+		})
+
+		return curseActions
+	}
+}
+
+func groupRMNSubjectBySelector(rmnSubjects []RMNCurseAction, filter bool) map[uint64][]Subject {
+	grouped := make(map[uint64][]Subject)
+	for _, subject := range rmnSubjects {
+		grouped[subject.ChainSelector] = append(grouped[subject.ChainSelector], subject.SubjectToCurse)
+	}
+
+	// Only keep unique subjects, preserve only global curse if present and eliminate any curse where the selector is the same as the subject
```

Comment:
> `eliminate any curse where the selector is the same as the subject`

Trying to understand this, aren't the subjects either global curse or the chain selector? So "eliminate any curse where the selector is the same as the subject" are just non-global curses no? Might be just simpler to say 'if the global curse subject is one of the subjects, prune all other subjects and only keep the global one'?

Reply:
I think my comment was not clear, I changed it, however the goal here was to avoid self curse or in other word having a chain curse it's own chain selector on RMNRemote